cluster/src/shard_lock_manager.rs

```diff
@@ -344,50 +491,104 @@ impl ShardLock {
         OnExpired: FnOnce(ShardId) -> Fut + Send + 'static,
         Fut: Future<Output = ()> + Send + 'static,
     {
-        let mut lease = Lease::new(lease_id, Duration::from_secs(self.ttl_sec));
-        let (lease_expire_notifier, mut lease_expire_receiver) = oneshot::channel();
-        let (keepalive_stop_sender, keepalive_stop_receiver) = oneshot::channel();
-        lease
-            .start_keepalive(
-                keepalive_stop_receiver,
-                lease_expire_notifier,
-                etcd_client,
-                runtime,
-            )
-            .await?;
-
-        let lock_expire_check_interval = LOCK_EXPIRE_CHECK_INTERVAL;
+        // Try to acquire the lock to ensure there is only one running keepalive
+        // procedure.
+        let initial_state = LeaseState::new(expired_at);
+        let lease = Arc::new(Lease::new(
+            lease_id,
+            Duration::from_secs(self.ttl_sec),
+            initial_state,
+        ));
+
         let shard_id = self.shard_id;
+        let rpc_timeout = self.rpc_timeout;
+        let lock_lease_check_interval = self.lease_check_interval;
+        let (keepalive_stop_sender, keepalive_stop_rx) = oneshot::channel();
+        let mut etcd_client = etcd_client.clone();
+        let runtime_for_bg = runtime.clone();
+        let lease_for_bg = lease.clone();
         let handle = runtime.spawn(async move {
+            let mut keepalive_stop_rx = keepalive_stop_rx;
             loop {
-                let timer = tokio::time::sleep(lock_expire_check_interval);
-                tokio::select! {
-                    _ = timer => {
-                        if lease.is_expired() {
-                            warn!("The lease of the shard lock is expired, shard_id:{shard_id}");
-                            on_lock_expired(shard_id);
-                            return
-                        }
+                if let Some(dur_until_expired)  = lease_for_bg.duration_until_expired() {
+                    // Assume the keepalive cost is 1.5 rpc_time to ensure the lease is not expired when waiting for next keepalive.
+                    let keepalive_cost = rpc_timeout + rpc_timeout / 2;
+                    if dur_until_expired > keepalive_cost {
+                        let wait_dur = dur_until_expired - keepalive_cost;
+                        info!(
+                            "Next keepalive will be schedule after {wait_dur:?}, shard_id:{shard_id}, lease_id:{lease_id}, \
+                            keepalive_cost:{keepalive_cost:?}, duration_until_expired:{dur_until_expired:?}"
+                        );
+                        tokio::time::sleep(wait_dur).await;
+
+                        info!("Start to keepalive, shard_id:{shard_id}, lease_id:{lease_id}");
+                    } else {
+                        warn!(
```

Comment:
Why do we need to expired in this case?

Reply:
There is no enough time to do a next keep alive call in the case where the timeout happens.